### ISSUE TITLE
remember last authentication method

### ISSUE THREAD

**Issue by johansmitsnl** [kanidm/kanidm#3328]
### Describe the solution you'd like

You can check to remember the username. In most cases I need to continue with the username already set, select passkey and again passkey and then your logged in. In 99% of the time I'll be using on the same device the same authentication method. Why not skip directly to the passkey authentication step if I've set it to remember it.

### Describe alternatives you've considered

No

**Comment by Firstyear** [comment]
This is a good suggestion, but it requires us to handle some new states in our authentication handlers that we currently can't. It's similar to #3218 

The tl;dr is our authentication flow internally is a series of challenge response steps, so if you "back" from a step (as you would need to do here), the state machine is in an invalid state and so the auth session moves to a failure. The idea was that you'd need to restart the flow to start again, but it makes sense to allow "back navigating" in general, so a way to "reset a partially completed flow" without actually resetting the whole thing would help here.

**Comment by yaleman** [comment]
We could detect/store the "preferred" mode client-side and send it as a signal to the begin step in a header or something, that'd be a relatively low impact way of doing it

**Comment by Firstyear** [comment]
If the user changes their mind and wants a different type of credential then we need to step back. That's the issue.

**Comment by yaleman** [comment]
Yep, but the signal is a request, the UI doesn't have to do it - if the hint option isn't in the list of options then prompt - and if they change their mind the way they tell Kanidm would be an interaction that clears the setting for next prompt

**Comment by Firstyear** [comment]
we have to submit the request to actually render the next prompt because when you pick a "mech" it returns with details like *challenges* that are needed for the request to proceed. we can't just do this lazy-maybe-render and only submit the mech choice on first interaction, because we actually need to advance the state machine to get back what we have to render.

**Comment by yaleman** [comment]
So when I type in my username and hit begin, when the browser sends the hint along-side that Kanidm can check if it's an option then the state machine can be advanced...?

**Comment by Firstyear** [comment]
Okay, and now if you don't want to use that mechanism? you have to go back. So the issue is the backward step, not the forward.

**Comment by yaleman** [comment]
So we need to support that option in the state machine, a different problem to https://github.com/kanidm/kanidm/issues/3328#issuecomment-2574438713

**Comment by johansmitsnl** [comment]
With for example Microsoft it uses by default passkey (in our tenant) but sometimes this is not available on the device I'm using. You can then cancel and select use other method where you are presented with a list of options.

Adding perhaps a "back" button that is part of the flow. Not the back button of the browser but an "use other" button.

**Comment by Firstyear** [comment]
> So we need to support that option in the state machine, a different problem to [#3328 (comment)](https://github.com/kanidm/kanidm/issues/3328#issuecomment-2574438713)

No, it's the same - if you hit the browser back button you're doing the same thing. I wanted to use mech type A, but now I want to step back to B. 

Just the same as here where the remember last mech used is fast-forwarding you to A, but then needs to allow stepping back to B (with a back button, or the browser back button). 

So yeah, it's pretty much the same issue, and then some other magic on top for the fastforward. But the majority of the work is in the auth state machine.
